pkg/cloud/openstack/clients/machineservice.go

```diff
@@ -290,7 +340,52 @@ func (is *InstanceService) InstanceCreate(name string, config *openstackconfigv1
 }
 
 func (is *InstanceService) InstanceDelete(id string) error {
-	return servers.Delete(is.computeClient, id).ExtractErr()
+	// get instance port id
+	allInterfaces, err := attachinterfaces.List(is.computeClient, id).AllPages()
+	if err != nil {
+		return err
+	}
+	instanceInterfaces, err := attachinterfaces.ExtractInterfaces(allInterfaces)
+	if err != nil {
+		return err
+	}
+	if len(instanceInterfaces) < 1 {
+		return fmt.Errorf("Failed to find the instance port")
+	}
+
+	// delete instance
+	err = servers.Delete(is.computeClient, id).ExtractErr()
+	if err != nil {
+		return err
+	}
+
+	// get and delete trunks
+	for _, port := range instanceInterfaces {
+		listOpts := trunks.ListOpts{
+			PortID: port.PortID,
+		}
+		allTrunks, err := trunks.List(is.networkClient, listOpts).AllPages()
+		if err != nil {
+			return err
```

Comment:
This code will error out if the installation of OpenStack does not support trunking... with the error "Resource not found". Even if trunks are not used to create VMs. Further cleanups will not happen.

In my test, I don't have an OpenStack with trunking, but I created a node (via machine object) and manually attached another device of a different subnet. Deletion of this node caused a left over port.

The error results in another reconciliation. But in the next loop, the machine is already gone, so no interfaces can be listed. Maybe it would be good to remove the VM at the very last step in this function. At least using the openstack cli tools, it is possible to delete ports that are attached to a running VM. However, trunking not being supported is not an error here.

Looping happens until
`I0114 15:47:26.492098    2447 actuator.go:195] Skipped deleting a VM that is already deleted.`

Reply:
I thought you could not delete the port of a VM (at least not the last one). Another option is checking if the VM exists, and only delete it if it is there, otherwise continue to remove the ports/trunks

Reply:
Umm, on a second thought... how will we not the associated ports... I'll check if all the ports can be deleted from a running VM

Reply:
ok, it seems you can delete all the ports... so VM should be deleted last! thanks!